src/liballoc/boxed.rs

```diff
@@ -523,7 +524,13 @@ impl<T: ?Sized> Box<T> {
     where
         T: 'a, // Technically not needed, but kept to be explicit.
     {
-        unsafe { &mut *Box::into_raw(b) }
+        let b = mem::ManuallyDrop::new(b);
+        let unique = b.0;
+        // Box is kind-of a library type, but recognized as a "unique pointer" by
+        // Stacked Borrows. Reborrowing with `&mut *` instead of transmuting
+        // ensures that the pointer we are returning here does not carry
+        // the tag of `b`, with `Unique` permission.
+        unsafe { &mut *unique.as_ptr() }
```

Comment:
Thinking about it... we are returning a mutable reference here. It shouldn't even be a problem to return something with `Unique` permission, quite the opposite, that is exactly what we want. So by going through `leak`, I think there is no way to not make Stacked Borrows happy.

How would you implement this method if you didn't have to worry about Stacked Borrows?

Reply:
Without (being aware of) Stacked Borrow I’d probably try to minimize the number of `unsafe` blocks first, number of "steps" second:

```rust
fn into_raw(b: Box<T>) -> *mut T { mem::ManuallyDrop::new(b).0.as_ptr() }
fn leak<'a>(b: Box<T>) -> &'a mut T { unsafe { &mut *Box::into_raw(b) }}
fn into_raw_non_null(b: Box<T>) -> NonNull<T> { Box::leak(b).into() }
```

Reply:
Ah I meant for implementing `Box::leak` directly, not going through `into_raw`.

Reply:
I’d inline the same: `unsafe { &mut *mem::ManuallyDrop::new(b).0.as_ptr() }` and would not have considered going through `Unique::as_mut` only to cast to a raw pointer and then borrow back to a reference.

Reply:
That code seems fine, actually. So maybe use that then with a comment like
```rust
        // Box is recognized as a "unique pointer" by Stacked Borrows, but internally it is a
        // raw pointer for the type system. Turning it directly into a raw pointer would not be
        // recognized as "releasing" the unique pointer to permit aliased raw accesses,
        // so all raw pointer methods go through `leak` which creates a (unique)
        // mutable reference. Turning *that* to a raw pointer behaves correctly.
```

Reply:
> So by going through `leak`, I think there is no way to not make Stacked Borrows happy.

Can you say more about this? What is the problem?

The premise of this PR is to recommend the `NonNull::from(Box::leak(b))` pattern, is that also problematic?

Reply:
There is no problem, I just used too many negations. ;)
"there is no way to *not* make Stacked Borrows happy", i.e., using `leak` will automatically make Stacked Borrows happy as we go through a reference explicitly.

Reply:
Reading more closely, that last code comment looks appropriate not for `leak` but for the *other* methods, the ones that call it. Does that sound right?